examples/misc/raptor.py

```diff
@@ -47,52 +77,58 @@ def task_state_cb(task, state):
 if __name__ == '__main__':
 
     if len(sys.argv) < 2:
-        cfg_file = './raptor.cfg'
+        cfg_file = '%s/%s' % (os.path.dirname(__file__), 'raptor.cfg')
     else:
         cfg_file = sys.argv[1]
 
-    cfg         = ru.Config(cfg=ru.read_json(cfg_file))
-    sleep       = int(cfg.sleep)
-    cpn         = cfg.cpn
-    gpn         = cfg.gpn
-    n_masters   = cfg.n_masters
-    n_workers   = cfg.n_workers
-    masters_pn  = cfg.masters_pn
-    nodes_pw    = cfg.nodes_pw
-    nodes_rp    = cfg.nodes_rp
-    workload    = cfg.workload
-    tasks_rp    = cfg.tasks_rp
-    nodes_agent = cfg.nodes_agent
-
-    # each master uses a node, and each worker on each master uses a node
-    # use 8 additional cores for non-raptor tasks
+    cfg              = ru.Config(cfg=ru.read_json(cfg_file))
+    sleep            = int(cfg.sleep)
+    cores_per_node   = cfg.cores_per_node
+    gpus_per_node    = cfg.gpus_per_node
+    n_masters        = cfg.n_masters
+    n_workers        = cfg.n_workers
+    masters_per_node = cfg.masters_per_node
+    nodes_per_worker = cfg.nodes_per_worker
+    nodes_rp         = cfg.nodes_rp
+    tasks_raptor     = cfg.tasks_raptor
+    tasks_rp         = cfg.tasks_rp
+    nodes_agent      = cfg.nodes_agent
+
+    cores_per_task   = cfg.cores_per_task
+
+    # Each master uses a node, and each worker also uses a node.
```

Comment:
The worker resources are allocated from the master resources, right? Or... somehow the master owns a pool of resources that it allocates to workers, doesn't it? Or does that not happen until the master actually submits workers (in which case, is it up to the user to be careful not to oversubscribe resources for workers)?

Please clarify whether the `cpu_threads` allocated to the `master` become unavailable to workers and their tasks.

Just to check: the scheduler does not trigger the Worker to pop a task unless the scheduler believes the Worker has enough idle threads to perform the task. Is that correct?